File: a2h/callbacks.py
"""
A2H webhook callbacks.

Instead of polling with human_check or blocking with gw.wait(),
register a callback that fires when a human responds.

    from a2h import Gateway, Participant
    from a2h.callbacks import CallbackRegistry

    gw = Gateway()
    callbacks = CallbackRegistry(gw)

    # Register a callback
    async def on_deal_approved(interaction):
        print(f"Deal approved: {interaction.response.value}")
        await execute_deal(interaction.context)

    callbacks.on_response("req_abc123", on_deal_approved)

    # Or register for ALL responses from a participant
    callbacks.on_any_response("sales/sarah", on_deal_approved)

For HTTP webhooks, use the server endpoint:

    POST /a2h/v1/webhooks
    {"url": "https://my-app.com/a2h-callback", "events": ["response", "expired"]}
"""

from __future__ import annotations

import asyncio
import logging
from typing import Any, Callable, Awaitable

from .models import Interaction, Status

logger = logging.getLogger(__name__)

Callback = Callable[[Interaction], Awaitable[None]]


class CallbackRegistry:
    """Register callbacks that fire when humans respond to A2H requests.

    Wraps the Gateway's respond method to trigger callbacks after
    a successful response is recorded.
    """

    def __init__(self, gateway):
        self._gw = gateway
        self._request_callbacks: dict[str, list[Callback]] = {}
        self._participant_callbacks: dict[str, list[Callback]] = {}
        self._global_callbacks: list[Callback] = []

        # Wrap the gateway's respond method
        self._original_respond = gateway.respond
        gateway.respond = self._intercepted_respond

    def _intercepted_respond(self, request_id: str, response_data: dict,
                              channel: str = "dashboard") -> dict:
        result = self._original_respond(request_id, response_data, channel)

        if result.get("success"):
            interaction = self._gw.get(request_id)
            if interaction:
                try:
                    loop = asyncio.get_running_loop()
                    loop.create_task(self._fire(interaction))
                except RuntimeError:
                    asyncio.run(self._fire(interaction))

        return result

    async def _fire(self, interaction: Interaction) -> None:
        """Fire all matching callbacks for this interaction."""
        callbacks: list[Callback] = []

        # Request-specific callbacks
        callbacks.extend(self._request_callbacks.pop(interaction.id, []))

        # Participant-specific callbacks
        pid = f"{interaction.to_namespace}/{interaction.to_name}"
        callbacks.extend(self._participant_callbacks.get(pid, []))

        # Global callbacks
        callbacks.extend(self._global_callbacks)

        for cb in callbacks:
            try:
                await cb(interaction)
            except Exception as e:
                logger.error("A2H callback failed: %s", e)

    # ---- Registration API --------------------------------------------------

    def on_response(self, request_id: str, callback: Callback) -> None:
        """Fire callback when a specific request gets a response.

        One-shot: the callback is removed after firing.
        """
        self._request_callbacks.setdefault(request_id, []).append(callback)

    def on_any_response(self, participant_pid: str, callback: Callback) -> None:
        """Fire callback when ANY request to this participant gets a response.

        Persistent: fires for every response, not just one.
        """
        self._participant_callbacks.setdefault(participant_pid, []).append(callback)

    def on_all_responses(self, callback: Callback) -> None:
        """Fire callback for every response in the system.

        Use for logging, metrics, or audit.
        """
        self._global_callbacks.append(callback)

    def remove(self, participant_pid: str | None = None) -> None:
        """Remove callbacks for a participant, or all if None."""
        if participant_pid:
            self._participant_callbacks.pop(participant_pid, None)
        else:
            self._request_callbacks.clear()
            self._participant_callbacks.clear()
            self._global_callbacks.clear()


class WebhookTarget:
    """Sends HTTP POST to a URL when events occur.

        target = WebhookTarget(
            url="https://my-app.com/a2h-callback",
            events=["response", "expired"],
            secret="shared-secret-for-hmac",
        )
        callbacks.on_all_responses(target.fire)
    """

    def __init__(self, url: str, events: list[str] | None = None, secret: str = ""):
        self.url = url
        self.events = events or ["response"]
        self.secret = secret

    async def fire(self, interaction: Interaction) -> None:
        event_type = "response" if interaction.status == Status.ANSWERED else interaction.status.value

        if event_type not in self.events:
            return

        payload = {
            "event": event_type,
            "interaction": interaction.to_dict(),
        }

        try:
            import httpx
            async with httpx.AsyncClient(timeout=10) as client:
                headers = {"Content-Type": "application/json"}
                if self.secret:
                    import hashlib, hmac, json
                    body = json.dumps(payload).encode()
                    sig = hmac.new(self.secret.encode(), body, hashlib.sha256).hexdigest()
                    headers["X-A2H-Signature"] = f"sha256={sig}"
                await client.post(self.url, json=payload, headers=headers)
        except Exception as e:
            logger.error("Webhook delivery failed to %s: %s", self.url, e)

File: src/typescript/callbacks.ts
import type { Interaction } from "./models.js";
import type { Gateway } from "./gateway.js";

type Callback = (interaction: Interaction) => Promise<void>;

export class CallbackRegistry {
  private _gateway: Gateway;
  private _originalRespond: Gateway["respond"];
  private _perRequest = new Map<string, Callback>();
  private _perParticipant = new Map<string, Callback>();
  private _global: Callback[] = [];

  constructor(gateway: Gateway) {
    this._gateway = gateway;
    this._originalRespond = gateway.respond.bind(gateway);
    gateway.respond = this._interceptedRespond.bind(this);
  }

  onResponse(requestId: string, callback: Callback): void {
    this._perRequest.set(requestId, callback);
  }

  onAnyResponse(participantPid: string, callback: Callback): void {
    this._perParticipant.set(participantPid, callback);
  }

  onAllResponses(callback: Callback): void {
    this._global.push(callback);
  }

  remove(participantPid?: string): void {
    if (participantPid) {
      this._perParticipant.delete(participantPid);
    } else {
      this._perRequest.clear();
      this._perParticipant.clear();
      this._global = [];
    }
  }

  reset(): void {
    this._gateway.respond = this._originalRespond;
    this.remove();
  }

  private _interceptedRespond(
    interactionId: string,
    responseData: Record<string, unknown>,
    channel = "dashboard"
  ): { success: boolean; status: string; error?: string } {
    const result = this._originalRespond(interactionId, responseData, channel);

    if (result.success) {
      const ix = this._gateway.get(interactionId);
      if (ix) {
        void this._fire(ix);
      }
    }

    return result;
  }

  private async _fire(interaction: Interaction): Promise<void> {
    // One-shot per request
    const reqCb = this._perRequest.get(interaction.id);
    if (reqCb) {
      this._perRequest.delete(interaction.id);
      try { await reqCb(interaction); } catch { /* ignore */ }
    }

    // Persistent per participant
    const pid = `${interaction.to_namespace}/${interaction.to_name}`;
    const partCb = this._perParticipant.get(pid);
    if (partCb) {
      try { await partCb(interaction); } catch { /* ignore */ }
    }

    // Global
    for (const cb of this._global) {
      try { await cb(interaction); } catch { /* ignore */ }
    }
  }
}

export class WebhookTarget {
  private _url: string;
  private _events: string[] | null;
  private _secret: string;

  constructor(url: string, opts?: { events?: string[]; secret?: string }) {
    this._url = url;
    this._events = opts?.events ?? null;
    this._secret = opts?.secret ?? "";
  }

  async fire(interaction: Interaction): Promise<void> {
    const eventType = interaction.response ? "response" : "expired";
    if (this._events && !this._events.includes(eventType)) return;

    const payload = JSON.stringify({
      event: eventType,
      interaction: interaction.toDict(),
    });

    const headers: Record<string, string> = {
      "Content-Type": "application/json",
    };

    if (this._secret) {
      const encoder = new TextEncoder();
      const key = await crypto.subtle.importKey(
        "raw",
        encoder.encode(this._secret),
        { name: "HMAC", hash: "SHA-256" },
        false,
        ["sign"]
      );
      const sig = await crypto.subtle.sign("HMAC", key, encoder.encode(payload));
      const hex = Array.from(new Uint8Array(sig))
        .map((b) => b.toString(16).padStart(2, "0"))
        .join("");
      headers["X-A2H-Signature"] = `sha256=${hex}`;
    }

    await fetch(this._url, { method: "POST", headers, body: payload });
  }
}
